Allow PP Ups to be edited

dot-Comfey opened the pull request:
Client-side change. I'm not a web designer so a lot of the HTML/CSS may be messy. However, it seems like the general interface works; PP ups can be edited, and they are able to properly import/export as well as pack/unpack. Pokemon sent into battle have the PP up changes applied.

![Screenshot 2024-02-10 165516](https://github.com/smogon/pokemon-showdown-client/assets/84290266/00dd86ab-c25a-4da6-84cd-16fe1ad71388)

dot-Comfey committed:
Allow PP Ups to be edited

Client-side change. I'm not a web developer so a lot of the HTML/CSS may be messy. However, it seems like the general interface works; PP ups can be edited, and they are able to properly import/export as well as pack/unpack.

dot-Comfey committed:
Less ugly CSS

dot-Comfey committed:
Moves and PP Ups on the same line

dot-Comfey commented:
![Screenshot 2024-02-11 201438](https://github.com/smogon/pokemon-showdown-client/assets/84290266/4eb9710f-0537-45eb-a1bc-52ee2b710d2a)

Karthik99999 commented:
> ![Screenshot 2024-02-11 201438](https://github.com/smogon/pokemon-showdown-client/assets/84290266/4eb9710f-0537-45eb-a1bc-52ee2b710d2a)
> 
> 

Just a suggestion, can we show the actual pp values in the dropdown instead of the amount of pp ups? I think it would be a lot more clear for non 5 pp moves. Not sure what others think about this though.

dot-Comfey committed:
Remove debugs

dot-Comfey commented:
They do default to max, and teams transferred to the new client should be able to unpack properly because moves without a defined PP up value get set to 3.

dot-Comfey committed:
Show move PP

dot-Comfey committed:
Update client-teambuilder.js

dot-Comfey commented:
![Screenshot 2024-02-15 234209](https://github.com/smogon/pokemon-showdown-client/assets/84290266/89e90e4a-14f9-47d3-872f-065dee236f21)

In the current implementation, PP ups will update every time an update to a move is made; PP ups are associated with a move slot and not a move, so if a move is changed, changes to PP ups will carry over, but if a move is removed from a set by trying to add a move twice, any changes to PP ups will continue to be associated with the original move.

DaWoblefet commented:
I'm not sure this should be editable in the UI at all, as opposed to being something that is just changeable via importing. We're taking up a lot of real estate in the teambuilder for a feature that will go unused the vast majority of the time.

dot-Comfey committed:
Remove UI

Only editable in the import/export interface now, which can be done by adding ';2', ';1', or ';0' after a move.

kwsch commented:
Based on my experience with editable entities, my thoughts:
1. Hide the PP selections by default.
2. Put a button-label on the far right above where it currently is (text=PP); italicize/color it if PP Ups deviate from the default max index. 
3. When the button is pressed, it will contract the move text boxes and show the PP Up overrides. Pressing it again will hide them and the moves will re-extend.
4. Italicize/color any selections that deviate from default (max).

For paste formatting, Moves are currently formatted as `- {move}`; possible inclusions for PP Ups:
1. Treat the `-` as the PP Up specifier; acknowledge `012-` as the first character to be interpreted as `0123` PP Ups.
2. Append an override after the move `(PP Ups: 1)`
3. Include a separate line if any PP is overridden before/after all moves (like `PP Ups: 3|1|3|3`)

dot-Comfey committed:
Merge branch 'smogon:master' into moveppups

dot-Comfey committed:
A little less ugly export formatting

Given that the original thread that suggested PP Up editing is still getting likes to this day, it's still a relevant suggestion and I wanted to revive this PR by making the export look less ugly. Not sure if we still want to have a UI in the teambuilder, as the PR thread suggests starting with changing the export format: https://www.smogon.com/forums/threads/editing-pp-on-showdown.3735890/

I'd also like to see if this works for server-side team storage, but I haven't figured out a way to test that.

KrisXV commented on play.pokemonshowdown.com/js/storage.js:
```diff
@@ -1493,7 +1526,9 @@ Storage.exportTeam = function (team, gen, hidestats) {
 			if (move) {
-				text += '- ' + move + "  \n";
+				text += '- ' + move;
+				if (curSet.movePPUps && curSet.movePPUps[j] < 3) text += " (PP Ups: " + curSet.movePPUps[j] + ")";
```
```suggestion
				if (curSet.movePPUps && curSet.movePPUps[j] && curSet.movePPUps[j] < 3) {
					text += " (PP Ups: " + curSet.movePPUps[j] + ")";
				}
```

KrisXV committed:
Update play.pokemonshowdown.com/js/storage.js

KrisXV commented on play.pokemonshowdown.com/js/storage.js:
```diff
@@ -913,10 +918,18 @@ Storage.fastUnpackTeam = function (buf) {
+		// move PP ups
+		if (buf.charAt(j) === ';') {
+			j = buf.indexOf('|', i);
+			set.movePPUps = buf.substring(i, j).split(',').map(number => parseInt(number));
```
```suggestion
			for (var q = 0; q < buf.substring(i, j).split(',').length; q++) {
				var ppUpsUsed = parseInt(buf.substring(i, j).split(',')[q]);
				set.movePPUps[q] = ppUpsUsed;
			}
```

larry-the-table-guy replied:
You might want to pull `buf.substring(i,j).split(,)` into a variable outside the for loop. Also `set.movePPUps` needs to be initialized w/ an array before writing to it.

dot-Comfey replied:
Initializing the array with the values as strings and converting them to numbers works without crashing or upsetting Linter, so I've done that.

dot-Comfey committed:
Update storage.js

dot-Comfey committed:
Fix unpacking

This should hopefully cover everything needed to implement PP Ups.

dot-Comfey committed:
Update panel-teamdropdown.tsx

dot-Comfey committed:
Change PP Ups display in packed format

PP Ups values that are 3 will now be shown as empty in the packed format, and if every value would be empty, PP Ups do not get displayed. I made this change to be consistent with how EVs and IVs are handled.

dot-Comfey committed:
Format errors

dot-Comfey committed:
Update storage.js

Zarel commented:
I would put this in the Details menu, rather than directly in the main window.

Zarel committed:
Merge branch 'master' into moveppups

Zarel committed:
Update for modern lint rules

dot-Comfey committed:
Add PP Ups to Details

dot-Comfey commented:
![Screenshot 2025-03-01 070457](https://github.com/user-attachments/assets/dfc028e4-5167-4045-a5b5-0ddb70206e64)

dot-Comfey committed:
Update client-teambuilder.js

dot-Comfey committed:
Trump Card exception

dot-Comfey committed:
Reset PP Ups if move is not Trump Card

I don't think people would want to realize after starting a battle that selecting Trump Card caused their move slot to have 0 PP Ups.

dot-Comfey committed:
Don't update PP Ups of other moves when selecting a new move

dot-Comfey committed:
Update unpack format

Updated on the server-side so doing the same here for consistency.

dot-Comfey committed:
Fix crash in unpacking

dot-Comfey committed:
Default Trump Card PP Ups to 0 everywhere

dot-Comfey committed:
Update storage.js
